### DISCUSSION TITLE
Fridges, Ovens: What is a typical use case

### DISCUSSION THREAD

**Discussion by QuickSander** [QuickSander/homebridge-mieleathome#15] [Ideas]
What information would you want to see from an oven or a fridge?

Oven:
- Target temperature
- Remaining time? (I have to verify if its is available from Miele as their API spec is off at several points)

Fridge:
- Current temperature only?

**Comment by MichelRabozee** [comment]
Hello Sander, seeing the possibilities of HomeKit, I do see only a few more possibilities (too bad there is no way to customise a label with the temperature sensor for instance; it would have allowed for the oven to put the program (grill, etc.) name besides the current temperature or remaining time.
Hence, here my proposals (maybe configurable through checkboxes to show or not some of the elements)
Oven:
- Target temperature
- Current temperature
- Remaining time
- ambientLight ("get"/"set" to switch it on)
- status : => powerOn/powerOff (possibly powerOff deactivable (like the "stop" for washer or dish washer))
- signalInfo
- signalFailure
- signalDoor

Fridge:
- Current temperature
- targetTemperature (get/set), "set" possibly deactivable (like the "stop" for washer or dish washer)
- signalDoor
- signalFailure
- signalInfo

**Reply by QuickSander** [reply]
I have to learn what is possible as for example the Miele API documentation is very off. It claims target temperature is not readable for my device, yet it is. 
It claims you can switch the light, yet I can't. 

Did you try any of the more challenging suggestions you made? Like setting the fridge target temperature? It would be "cool" but I have a hard time believing it is possible after seeing what my washing machine mostly cannot do

**Comment by MichelRabozee** [comment]
I will try to test my proposals this afternoon, you are right, they are only coming from the swagger system, I was not aware it is not "up to date" :-(

I come back to you.

**Comment by MichelRabozee** [comment]
testing Fridge:

Values retrieved from "state":
https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/state?language=en

OK temperature (raw: °C * 100, localized: °C)
OK targetTemperature (raw: °C * 100, localized: °C)
OK signalFailure
OK signalInfo
OK signalDoor (true if open)

cf.:
{
  "ProgramID": {
    "value_raw": 2466250752,
    "value_localized": "",
    "key_localized": "Program name"
  },
  "status": {
    "value_raw": 5,
    "value_localized": "In use",
    "key_localized": "status"
  },
  "programType": {
    "value_raw": 0,
    "value_localized": "Programa",
    "key_localized": "Program type"
  },
  "programPhase": {
    "value_raw": 0,
    "value_localized": "",
    "key_localized": "Program phase"
  },
  "remainingTime": [
    0,
    0
  ],
  "startTime": [
    0,
    0
  ],
  "targetTemperature": [
    {
      "value_raw": 400,
      "value_localized": 4,
      "unit": "Celsius"
    }
  ],
  "temperature": [
    {
      "value_raw": 400,
      "value_localized": 4,
      "unit": "Celsius"
    }
  ],
  "signalInfo": false,
  "signalFailure": false,
  "signalDoor": false,
  "remoteEnable": {
    "fullRemoteControl": true,
    "smartGrid": false,
    "mobileStart": false
  },
  "ambientLight": 2,
  "light": null,
  "elapsedTime": [],
  "spinningSpeed": {
    "unit": "rpm",
    "value_raw": null,
    "value_localized": null,
    "key_localized": "Spin speed"
  },
  "dryingStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Drying level"
  },
  "ventilationStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Fan level"
  },
  "plateStep": [],
  "ecoFeedback": null,
  "batteryLevel": null
}



Actions: from "actions":

SET targetTemperature, "set" possibly deactivable (like the "stop" for washer or dish washer) in  °C
curl -X PUT "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: */*" -H "Authorization: Bearer BE_xxx" -H "Content-Type: application/json" -d "{\"targetTemperature\":[{\"zone\":1,\"value\":4}]}"
==>	works even if system returns "error 500" instead of "OK 204"


"DynaCool": processAction 6 = ON, processAction 7 = OFF
ON:
curl -X PUT "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: */*" -H "Authorization: Bearer BE_xxx" -H "Content-Type: application/json" -d "{\"processAction\":6}"
Status returns "Super cool" ???:
   "status": {
    "value_raw": 14,
    "value_localized": "Super cool",
    "key_localized": "status"
  },

OFF:
curl -X PUT "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: */*" -H "Authorization: Bearer BE_xxx" -H "Content-Type: application/json" -d "{\"processAction\":7}"


cf.:
curl -X GET "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: application/json" -H "Authorization: Bearer BE_xxx"
{
  "processAction": [
    6
  ],
  "light": [],
  "ambientLight": [
    1
  ],
  "startTime": [],
  "ventilationStep": [],
  "programId": [],
  "targetTemperature": [
    {
      "zone": 1,
      "min": 1,
      "max": 9
    }
  ],
  "deviceName": true,
  "powerOn": false,
  "powerOff": false,
  "colors": [],
  "modes": []
}

Next to test: oven :-)

**Reply by QuickSander** [reply]
> SET targetTemperature

Do you happen to know the valid temperature range? Then I can set the correct properties in Homekit for the temperature dial. Around 0-10 ?  Where '0' will probably return an error I guess.

**Reply by MichelRabozee** [reply]
According to possible "actions", range is 1 to 9, zone is 1

https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions
{
  "processAction": [
    6
  ],
  "light": [],
  "ambientLight": [
    1
  ],
  "startTime": [],
  "ventilationStep": [],
  "programId": [],
  "targetTemperature": [
    {
      "zone": 1,
      "min": 1,
      "max": 9
    }
  ],
  "deviceName": true,
  "powerOn": false,
  "powerOff": false,
  "colors": [],
  "modes": []
}

**Reply by MichelRabozee** [reply]
curl -X PUT "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: */*" -H "Authorization: Bearer BE_XXX" -H "Content-Type: application/json" -d "{\"targetTemperature\":[{\"zone\":1,\"value\":4}]}"

It is important to note that the answer is error 500
{
  "code": 500,
  "message": "A technical error occured. Please try again later. (Error-ID: d1c2771c1f8ccf82)"
}

BUT: the command went through, the new target temperature is taken into account !!!

**Reply by QuickSander** [reply]
That error code triggers me, as I noticed another person having trouble logging into the Miele API website which returned also with error code 500 with the same text (although it is very likely they use the same error text for all 500 errors).
So it might be temporary, but I will keep it in mind when implementing.

**Reply by MichelRabozee** [reply]
The 500 error in HTTP protocol means that the server thread crashed. It can be a number of things, generally a bad server coding not coping with some situation. I do not know about the case you mention (and in that case, it may be due to a bad entry or whatever... seem the Miele server is not very resilient) but I guess in the case of the targetTemperature setting, the HTTP server thread crashes after having correctly processed the request (I can see in the Miele app (where you can change the value as well) and on the refrigerator display (after it reached the new temperature) that the target temperature has changed to the value I want).

**Reply by QuickSander** [reply]
I wouldn't say the HTTP 500 error code means the server thread crashed. It is not defined so specifically as such in the HTTP standard. It can be anything. It is up to Miele to decide, but it generally means _any_ internal server error. 

It just triggered me that yesterday another person encountered the same error for a HTTP request which was well formed and did work before, so it might be temporary. But who knows. It might be very well a bug in the processing code.

**Reply by MichelRabozee** [reply]
(your comment disappeared while I was answering :-( )

The "zone" must always be "1", and the value must be between 1 and 9
So to set targetTemperature to 4:
curl -X PUT "https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions" -H "accept: /" -H "Authorization: Bearer BE_xxx" -H "Content-Type: application/json" -d **"{"targetTemperature":[{"zone":1,"value":4}]}"**

**Reply by QuickSander** [reply]
Yes because I realized the same... 😁

If there are more zones, I could potentially generate a control per zone.

**Reply by QuickSander** [reply]
I've published a new version with very experimental fridge support. 
I do expect some bugs as I am unable to test the fridge myself fully.

E.g. I could not reproduce the status 500 reply (for me it replies 400 / Bad request while the content type and request seem well formed to me, so it might be just my device reacting like this which sounds logical to me)

**Reply by MichelRabozee** [reply]
Hello, nice ^^

I have no major issue, display of temperature is OK, setting the target temperature is OK as well; of course, I get the 500 error in the logs from homebridge:
```
[06/03/2021 12:22:45] [Miele] Miele API request https://api.mcs3.miele.com/v1/devices/mac-00124B0005AB4D8B/actions failed. Status 500: "Internal Server Error".Error: Request failed with status code 500
```
One strange thing: when I set the new target temperature to 3° (instead of previous 4°), the device state from the Miele API says:
```
  "targetTemperature": [
    {
      "value_raw": 300,
      "value_localized": 3,
      "unit": "Celsius"
    }
  ],
  "temperature": [
    {
      "value_raw": 400,
      "value_localized": 4,
      "unit": "Celsius"
    }
  ],
```
so the target temperature is not reached yet (The Fridge display is still indicating 4°). But if the Home display (and in the Homebridge Accessory display), I already get a 3.1 (or 3.0 in Home) value... as if the value from the target temperature instantly supersedes the current temperature in the plugin (in the homebridge accessory display page, you see that if you do refresh the page while the target temperature is still not reached, the current value appears briefly, then the target value is displayed - in the circle on top I mean, where the current temperature is shown; in the bottom value, I guess it is the target temperature which is displayed, and that is always correct and steady -). Of course, when target temperature is reached, all values are correct.

I also got 2 warnings at homebridge starting:
```
[06/03/2021 12:20:14] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': characteristic was supplied illegal value: number 0 exceeded minimum of 1. See https://git.io/JtMGR for more info.
[06/03/2021 12:20:15] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Status Fault': Characteristic not in required or optional characteristic section for service Thermostat. Adding anyway.. See https://git.io/JtMGR for more info.
```
The first one is maybe normal (allowed range for Miele fridge is 1-9 degrees, however, at startup, it seems the characteristic is initialised to 0. Not sure what to change (init value or range) to avoid that...
The second warning, I do not know.
Anyway, those are warnings and somewhat appearing only at startup...

**Reply by QuickSander** [reply]
> I get the 500 error in the logs from homebridge

I figured to log the error instead of silently ignoring it. I will extend it to log it as a warning instead in case of exactly error code 500.

> as if the value from the target temperature instantly supersedes the current temperature

I had a mistake in the generic update code to always update the current temperature. Will fix this.

```
[06/03/2021 12:20:14] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': characteristic was supplied illegal value: number 0 exceeded minimum of 1. See https://git.io/JtMGR for more info.
[06/03/2021 12:20:15] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Status Fault': Characteristic not in required or optional characteristic section for service Thermostat. Adding anyway.. See https://git.io/JtMGR for more info.
```

These two I have to think about what to do with them as I added the GeneralFault Characteristic on purpose to make an error visible in Eve app instead of just the Homebridge log.

**Reply by MichelRabozee** [reply]
> > I get the 500 error in the logs from homebridge
> 
> I figured to log the error instead of silently ignoring it. I will extend it to log it as a warning instead in case of exactly error code 500.
> 
OK, but only for that functionality (setting the target temperature for a fridge). In other cases, the error could be a real error :-)

I will check your new release (I missed the 2.6.1, you were too fast releasing 2.7.0 :-) )

**Reply by MichelRabozee** [reply]
Just to mention that now I got some new warnings appearing, the last three ones repeating every minute):
```
[06/03/2021 19:56:11] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': Characteristic not in required or optional characteristic section for service Service. Adding anyway.. See https://git.io/JtMGR for more info.
[06/03/2021 19:56:11] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': characteristic was supplied illegal value: number 1 exceeded minimum of 10. See https://git.io/JtMGR for more info.
[06/03/2021 19:57:10] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': characteristic was supplied illegal value: number 1 exceeded minimum of 10. See https://git.io/JtMGR for more info.
[06/03/2021 19:58:10] [homebridge-mieleathome] This plugin generated a warning from the characteristic 'Target Temperature': characteristic was supplied illegal value: number 1 exceeded minimum of 10. See https://git.io/JtMGR for more info.
```
I am puzzled with the first one, regarding a service "Service" ???

BTW, the display of current temperature vs target one is fixed; and the warning for error 500 is perfect ! ;-) ^^

**Reply by QuickSander** [reply]
I'm afraid, I cannot think of much to solve this warning. I'm bound to Homebridge and HomeKit. HomeKit's minimum target temperature is 10°C. 

I could try to not update it that often to reduce the amount of warnings. But I think it is something we have to live with. The thermostat is simply not designed for fridges.

**Reply by MichelRabozee** [reply]
The thing is that this recurrent warning (the one about the target temperature which must be over 10) was not present with version 2.6.0...
Maybe this is a side-effect of the first warning I mentioned "'Target Temperature': Characteristic not in required or optional characteristic section for service Service."... That warning about an unlisted  "Target Temperature" characteristic was not present in 2.6.0

**Reply by QuickSander** [reply]
I understand, but it was the bug that did not update the target temperature before. That's why you didn't see it. I will at least reduce the frequency (no update when value is the same)

**Reply by QuickSander** [reply]
See: https://developers.homebridge.io/#/characteristic/TargetTemperature

The service thing I have to investigate since it almost seems like a homebridge bug. Target temp is namely a valid characteristic for a thermostat.

**Reply by MichelRabozee** [reply]
> I understand, but it was the bug that did not update the target temperature before. That's why you didn't see it. I will at least reduce the frequency (no update when value is the same)

Maybe it is possible to change the acceptable range of the TargetTemperature characteristic:
```
		<thermostat service>
			.getCharacteristic(Characteristic.TargetTemperature)
			.setProps({
						maxValue: 9,
						minValue: 1,
						minStep: 1.00
					});
```

**Reply by MichelRabozee** [reply]
> See: https://developers.homebridge.io/#/characteristic/TargetTemperature
> 
> The service thing I have to investigate since it almost seems like a homebridge bug. Target temp is namely a valid characteristic for a thermostat.

My point was that the warning speaks about a service "Service" instead of a service "Thermostat", as if the TargetTemperature characteristic is not added to a "Thermostat" but rather to a generic "Service".

**Reply by QuickSander** [reply]
> Maybe it is possible to change the acceptable range of the TargetTemperature characteristic:

I do not understand what you mean to suggest. This is namely exactly what I do already (otherwise you wouldn't have seen the dial starting at 1 and ending at 9 with steps of 1 degree). The point is that for a normal "current" temperature the range is absolute 0 Kelvin to an acceptable range for earth temperatures. The target temperature is however starting at 10 degrees. Which is logical if you consider the fact it is designed for thermostats. It is not something you can override. The range you set with the syntax you specified should always be within the HomeKit range. You can only narrow it, not make it broader.

> My point was that the warning speaks about a service "Service"

I understand, but I recognise this generic "service" message. I need some time to figure this out, can be as well the status fault I add since that is certainly not an optional characteristic for a thermostat service.

**Reply by MichelRabozee** [reply]
OK, I understand (sorry, I did not go through the code to look first :-( ). I had a similar boundary issue with a Humidity sensor which I use as a Water Level, and tough normally it is limited to 100, I succeeded to change the upper boundary to 125 (for instance, when the water level is a bit above the normal height, the sensor may indicate 105). Now, there are no more warning in Homebridge. Maybe the relative humidity characteristic is more prone to changing boundaries. Anyway, your way (only notifying Homekit if the value changes) is perfect.

**Reply by MichelRabozee** [reply]
I installed 2.7.1, and the plugin is much less verbose :-) In fact, no more warnings at all regarding the acceptable range of the target temperature and also the warning about that characteristic being added to the "Service" service disappeared as well!!!

Thank you, you are a chief!

**Reply by QuickSander** [reply]
Thanks. I hope this remains with v2.7.2. Since if it doesn't and reappears I might have a good idea what I did wrong.

**Reply by MichelRabozee** [reply]
2.7.2 fixes the temperature for the dish washer and wash machine, and still no Warning or Error at startup !

--EDIT--: seemed there was a communication problem with the Miele servers, which was seeing as all my appliances being "Disconnected".

Thus, disregards what I wrote as side effects if you already read my post which I in the meantime edited :-)

**Reply by QuickSander** [reply]
Oké thanks. I was puzzled there for a moment. :-D 
I do start to wonder though how the plugin should have behaved in case of connection problems.

**Reply by MichelRabozee** [reply]
So, I had to restart my Gateway. I seldom arrives (twice in 3 years I think). No clue what happened. All was locally OK (as seen by the gateway), but in the app and in the API (devices and short/devices), the appliances were seen as "disconnected". The "state" API however behaved strangely for the Wash machine, it returned a 404 error at HTTP level. The Fridge and Dish Washer returned a correct JSON with a status 255 (if I recall correctly, I should have saved the JSON :-( ) - disconnected.
The plugin reacted in weird way, effectvely (showing wrong ON/OFF state or target temperature OFF). I do not know if something can be done for that.

**Comment by MichelRabozee** [comment]
Oven: no action possible as you guessed...

OK targetTemperature
OK temperature
OK remainingTime
OK status : => powerOn/powerOff (possibly powerOff deactivable (like the "stop" for washer or dish washer))
OK signalInfo
OK signalFailure
OK signalDoor (true if open)

{
  "ProgramID": {
    "value_raw": 7,
    "value_localized": "Fan plus",
    "key_localized": "Program name"
  },
  "status": {
    "value_raw": 6,
    "value_localized": "Pause",
    "key_localized": "status"
  },
  "programType": {
    "value_raw": 0,
    "value_localized": "Programa",
    "key_localized": "Program type"
  },
  "programPhase": {
    "value_raw": 3072,
    "value_localized": "Cooling phase",
    "key_localized": "Program phase"
  },
  "remainingTime": [
    0,
    0
  ],
  "startTime": [
    0,
    0
  ],
  "targetTemperature": [
    {
      "value_raw": 16000,
      "value_localized": 160,
      "unit": "Celsius"
    },
    {
      "value_raw": -32768,
      "value_localized": null,
      "unit": "Celsius"
    }
  ],
  "temperature": [
    {
      "value_raw": 5200,
      "value_localized": 52,
      "unit": "Celsius"
    },
    {
      "value_raw": -32768,
      "value_localized": null,
      "unit": "Celsius"
    }
  ],
  "signalInfo": false,
  "signalFailure": false,
  "signalDoor": true,
  "remoteEnable": {
    "fullRemoteControl": false,
    "smartGrid": false,
    "mobileStart": false
  },
  "ambientLight": null,
  "light": null,
  "elapsedTime": [
    0,
    1
  ],
  "spinningSpeed": {
    "unit": "rpm",
    "value_raw": null,
    "value_localized": null,
    "key_localized": "Spin speed"
  },
  "dryingStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Drying level"
  },
  "ventilationStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Fan level"
  },
  "plateStep": [],
  "ecoFeedback": null,
  "batteryLevel": null
}

**Actions: even if some are listed, none have effect (PowerOn, powerOff, ambientLight)**

https://api.mcs3.miele.com/v1/devices/mac-001D63FFFE0203B2/actions
{
  "processAction": [],
  "light": [],
  "ambientLight": [
    1
  ],
  "startTime": [],
  "ventilationStep": [],
  "programId": [],
  "targetTemperature": [],
  "deviceName": false,
  "powerOn": true,
  "powerOff": false,
  "colors": [],
  "modes": []
}
